src/ngraph_assign_clusters.cc

```diff
@@ -493,14 +550,51 @@ Status AssignClusters(Graph* graph) {
         MergeClusters(edge, cluster_map);
         // something changed
         changed = true;
+      } else {
+        if (collect_non_contracting_edge_info) {
+          // either static input
+          // or there exists a longer path, so contracting this edge causes
+          // cycles
+          std::vector<int32> static_inputs;
+          GetStaticInputs(dst, &static_inputs);
+          bool is_static = std::find(static_inputs.begin(), static_inputs.end(),
+                                     edge->dst_input()) != static_inputs.end();
+          // 3 possible reasons here:
+          // src dst lies in same cluster, so nothing to do (trivial cycle
+          // induced in graphcycles)
+          // dst has static input
+          // a longer irreducible path exists
+          cluster_separation_reason[get_string_key(src_index, dst_index)]
+              .push_back(src_index == dst_index
+                             ? EdgeNonContractionReasons::SAMECLUSTER
+                             : (is_static
+                                    ? EdgeNonContractionReasons::STATICINPUT
+                                    : EdgeNonContractionReasons::PATHEXISTS));
+        }
+      }
+    }
+
+    if (!changed && config::IsLoggingPlacement()) {
+      // This will be entered only once if logging is enabled
+      // When entered, it will force the do-while to run one last time,
+      // collecting information
+      if (!collect_non_contracting_edge_info) {
+        changed = true;
+        collect_non_contracting_edge_info = true;
+      } else {
+        if (changed) {
```

Comment:
the outer if loop checks for !changed so is the else -> if(changed) code block required?

Reply:
This if does not do anything. It just asserts, "you should NOT be in this state"

ie, if collect_non_contracting_edge_info is true, then changed should be false (else raise error)

Reply:
changed will be always false, as the outer if loop checks if(!changed)

Reply:
ah yes. I see